projects/pswg/src/main/java/com/parzivail/pswg/entity/ThermalDetonatorEntity.java

```diff
@@ -0,0 +1,135 @@
+import net.minecraft.registry.tag.BlockTags;
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.server.world.ServerWorld;
+import net.minecraft.util.hit.BlockHitResult;
+import net.minecraft.util.hit.HitResult;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.world.BlockView;
+import net.minecraft.world.World;
+import net.minecraft.world.explosion.Explosion;
+
+public class ThermalDetonatorEntity extends ThrowableExplosive implements IPrecisionSpawnEntity, IPrecisionVelocityEntity
+{
+	public ThermalDetonatorEntity(EntityType<ThermalDetonatorEntity> type, World world)
+	{
+		super(type, world);
+		setExplosionPower(5f);
+	}
+
+	@Override
+	public void tick() {
+
+		this.speed = this.speed*0.95f;
+		super.tick();
+	}
+
+	@Override
+	protected void createParticles(ServerWorld world, double x, double y, double z){
+		int m= (int)explosionPower/2;
+		for (int i = 0; i < world.getPlayers().size(); ++i) {
+			ServerPlayerEntity player =world.getPlayers().get(i);
+			world.spawnParticles(player, ParticleTypes.FLASH, true, x, y, z, 1, 0, 0,0, 0);
+
+			world.spawnParticles(player,ParticleTypes.LARGE_SMOKE, true, x, y, z, 50*m, 1, 1, 1, 0);
+			world.spawnParticles(player,ParticleTypes.SMOKE, true, x, y, z, 45*m, 0.75*m, 0.75*m, 0.75*m, 0);
+
+			world.spawnParticles(player,SwgParticles.EXPLOSION_SMOKE, true, x, y+0.5f, z, 15*m, 0.125*m, 0.25*m, 0.125*m, 0.03);
+			world.spawnParticles(player,SwgParticles.EXPLOSION_SMOKE, true, x, y+0.5f, z, 15*m, 0.25*m, 0.125*m, 0.125*m, 0.025);
+			world.spawnParticles(player,SwgParticles.EXPLOSION_SMOKE, true, x, y+0.5f, z, 15*m, 0.125*m, 0.125*m, 0.25*m, 0.02);
+			world.spawnParticles(player,SwgParticles.EXPLOSION_SMOKE, true, x, y+0.5f, z, 10*m, 0.1*m, 0.2*m, 0.1*m, 0.0075);
+			world.spawnParticles(player,SwgParticles.EXPLOSION_SMOKE, true, x, y+0.5f, z, 10*m, 0.2*m, 0.1*m, 0.25*m, 0.01);
+			world.spawnParticles(player, SwgParticles.EXPLOSION_SMOKE, true, x, y+0.5f, z, 10 * m, 0.25 * m , 0.5 * m, 0.3 * m, 0.015);
+
+			world.spawnParticles(player,ParticleTypes.FLAME, true, x, y, z, 20*m, 0.5*m, 0.5*m, 0.5*m, 0.1);
+			world.spawnParticles(player,ParticleTypes.SMALL_FLAME, true, x, y, z, 20*m, 1.2*m, 1.2*m, 1.2*m, 0.125);
+			world.spawnParticles(player,ParticleTypes.SMALL_FLAME, true, x, y, z, 20*m, 1.1*m, 1.3*m, 1.25*m, 0.125);
+		}
+	}
+
+	@Override
+	public boolean canBeHitByProjectile()
+	{
+		return true;
+	}
+
+	@Override
+	protected void onCollision(HitResult hitResult)
+	{
+
+		if (hitResult.getType() == HitResult.Type.BLOCK)
+		{
+			if(hitResult.squaredDistanceTo(this)<0.01){
+				this.setVelocity(0,0,0);
+			}
+			BlockHitResult blockHitResult = (BlockHitResult)hitResult;
+
+			Vec3d velocity = this.getVelocity();
+			Vec3d pos = hitResult.getPos();
+
+			double modX = 0.9f;
+			double modY = 0.9f;
+			double modZ = 0.9f;
+			if(blockHitResult.getSide()==Direction.UP){
```

Comment:
I think you can do something like `var normal = blockHitResult.getSide().getNormal()` and then you can reflect `velocity` about that normal vector using `velocity = MathUtil.reflect(velocity, normal);` to prevent this huge `if` chain

Reply:
"blockHitResult.getSide().getNormal()" isn't a method, and I'm not exactly sure how to get it?

Reply:
I mean `Vec3d pos = hitResult.getPos();` exists, I guess I'll try that?

Reply:
Yeah, no, it has the completly wrong effect.

Reply:
It just reverses the vectors instead of actually determining what needs to be inversed